src/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.Runtime.CompilerHelpers
+{
+    /// <summary>
+    /// Array helpers for generated code.
+    /// </summary>
+    internal static class ArrayHelpers
+    {
+        /// <summary>
+        /// Helper for array allocations via `newobj` IL instruction. Dimensions are passed in as block of integers.
+        /// </summary>
+        private static unsafe Array NewObjArray(IntPtr pEEType, int nDimensions, int* pDimensions)
+        {
+            EETypePtr eeType = new EETypePtr(pEEType);
+            Debug.Assert(eeType.IsArray);
+            Debug.Assert(nDimensions > 0);
+
+            if (eeType.IsSzArray)
+            {
+                Array ret = (Array)RuntimeImports.RhNewArray(eeType, pDimensions[0]);
+
+                if (nDimensions > 1)
+                {
+                    // Allocate jagged array
+                    EETypePtr elementType = eeType.ArrayElementType;
+                    Debug.Assert(elementType.IsSzArray);
+
+                    Array[] arrayOfArrays = (Array[])ret;
+                    for (int i = 0; i < arrayOfArrays.Length; i++)
+                        arrayOfArrays[i] = NewObjArray(elementType.RawValue, nDimensions - 1, pDimensions + 1);
+                }
+
+                return ret;
+            }
+
+            int rank = eeType.ArrayRank;
+            Debug.Assert(rank == nDimensions || 2 * rank == nDimensions);
+
+            if (rank < nDimensions)
+            {
+                for (int i = 0; i < rank; i++)
+                {
+                    if (pDimensions[2 * i] != 0)
+                        throw new PlatformNotSupportedException(SR.Arg_NotSupportedNonZeroLowerBound);
+
+                    pDimensions[i] = pDimensions[2 * i + 1];
```

Comment:
"Modifies memory pointed to by `pDimensions`" might be worth calling out in the summary